### PR TITLE
gperf: conan v2 support

### PR DESCRIPTION
**Pull request by SpaceIm**
Specify library name and version:  **lib/1.0**

This is also a good place to share with all of us **why you are submitting this PR** (specially if it is a new addition to ConanCenter): is it a dependency of other libraries you want to package? Are you the author of the library? Thanks!

---

- [ ] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [ ] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [ ] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [ ] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by SpaceIm** [comment]
```
/c/j/w/prod/buildsinglereference/.conan/data/gperf/3.1/_/_/build/01edd76db8e16db9b38c3cca44ec466a9444c388/src/build-aux/compile cl -nologo -MTd -Zi -Ob0 -Od -FS -D_WIN32_WINNT=_WIN32_WINNT_WIN8 -I/c/j/w/prod/buildsinglereference/.conan/data/gperf/3.1/_/_/build/01edd76db8e16db9b38c3cca44ec466a9444c388/src/lib -c /c/j/w/prod/buildsinglereference/.conan/data/gperf/3.1/_/_/build/01edd76db8e16db9b38c3cca44ec466a9444c388/src/lib/getopt.c
/c/j/w/prod/buildsinglereference/.conan/data/gperf/3.1/_/_/build/01edd76db8e16db9b38c3cca44ec466a9444c388/src/build-aux/compile cl -nologo -MTd -Zi -Ob0 -Od -FS -D_WIN32_WINNT=_WIN32_WINNT_WIN8 -I/c/j/w/prod/buildsinglereference/.conan/data/gperf/3.1/_/_/build/01edd76db8e16db9b38c3cca44ec466a9444c388/src/lib -c /c/j/w/prod/buildsinglereference/.conan/data/gperf/3.1/_/_/build/01edd76db8e16db9b38c3cca44ec466a9444c388/src/lib/getopt1.c
/c/j/w/prod/buildsinglereference/.conan/data/gperf/3.1/_/_/build/01edd76db8e16db9b38c3cca44ec466a9444c388/src/build-aux/compile cl -nologo -MTd -Zi -Ob0 -Od -FS -D_WIN32_WINNT=_WIN32_WINNT_WIN8 -I/c/j/w/prod/buildsinglereference/.conan/data/gperf/3.1/_/_/build/01edd76db8e16db9b38c3cca44ec466a9444c388/src/lib -c /c/j/w/prod/buildsinglereference/.conan/data/gperf/3.1/_/_/build/01edd76db8e16db9b38c3cca44ec466a9444c388/src/lib/getline.cc
/c/j/w/prod/buildsinglereference/.conan/data/gperf/3.1/_/_/build/01edd76db8e16db9b38c3cca44ec466a9444c388/src/build-aux/compile cl -nologo -MTd -Zi -Ob0 -Od -FS -D_WIN32_WINNT=_WIN32_WINNT_WIN8 -I/c/j/w/prod/buildsinglereference/.conan/data/gperf/3.1/_/_/build/01edd76db8e16db9b38c3cca44ec466a9444c388/src/lib -c /c/j/w/prod/buildsinglereference/.conan/data/gperf/3.1/_/_/build/01edd76db8e16db9b38c3cca44ec466a9444c388/src/lib/hash.cc
getopt.c
getopt1.c
hash.cc
getline.cc
c1xx: fatal error C1083: Cannot open source file: 'C:J:/w/prod/buildsinglereference/.conan/data/gperf/3.1/_/_/build/01edd76db8e16db9b38c3cca44ec466a9444c388/src/lib/hash.cc': No such file or directory
c1xx: fatal error C1083: Cannot open source file: 'C:J:/w/prod/buildsinglereference/.conan/data/gperf/3.1/_/_/build/01edd76db8e16db9b38c3cca44ec466a9444c388/src/lib/getline.cc': No such file or directory
make[1]: Leaving directory '/c/J/w/prod/BuildSingleReference/.conan/data/gperf/3.1/_/_/build/01edd76db8e16db9b38c3cca44ec466a9444c388/build-debug/lib'
```

mysteries of compilers...

**Comment by SpaceIm** [comment]
Seems to be quite similar than error in https://github.com/conan-io/conan-center-index/pull/13446

**Comment by uilianries** [comment]
I just started a new CI build right now.

**Comment by SpaceIm** [comment]
Ok thanks. I still don't undestand why C++ files are not found, but C files are.

**Comment by jcar87** [comment]
These 4 commands seem correct:
```
/c/j/w/prod/buildsinglereference/.conan/data/gperf/3.1/_/_/build/0a420ff5c47119e668867cdb51baff0eca1fdb68/src/build-aux/compile cl -nologo -MT -O2 -Ob2 -FS -DNDEBUG -D_WIN32_WINNT=_WIN32_WINNT_WIN8 -I/c/j/w/prod/buildsinglereference/.conan/data/gperf/3.1/_/_/build/0a420ff5c47119e668867cdb51baff0eca1fdb68/src/lib -c /c/j/w/prod/buildsinglereference/.conan/data/gperf/3.1/_/_/build/0a420ff5c47119e668867cdb51baff0eca1fdb68/src/lib/getopt.c
/c/j/w/prod/buildsinglereference/.conan/data/gperf/3.1/_/_/build/0a420ff5c47119e668867cdb51baff0eca1fdb68/src/build-aux/compile cl -nologo -MT -O2 -Ob2 -FS -DNDEBUG -D_WIN32_WINNT=_WIN32_WINNT_WIN8 -I/c/j/w/prod/buildsinglereference/.conan/data/gperf/3.1/_/_/build/0a420ff5c47119e668867cdb51baff0eca1fdb68/src/lib -c /c/j/w/prod/buildsinglereference/.conan/data/gperf/3.1/_/_/build/0a420ff5c47119e668867cdb51baff0eca1fdb68/src/lib/getopt1.c
/c/j/w/prod/buildsinglereference/.conan/data/gperf/3.1/_/_/build/0a420ff5c47119e668867cdb51baff0eca1fdb68/src/build-aux/compile cl -nologo -MT -O2 -Ob2 -FS -DNDEBUG -D_WIN32_WINNT=_WIN32_WINNT_WIN8 -I/c/j/w/prod/buildsinglereference/.conan/data/gperf/3.1/_/_/build/0a420ff5c47119e668867cdb51baff0eca1fdb68/src/lib -c /c/j/w/prod/buildsinglereference/.conan/data/gperf/3.1/_/_/build/0a420ff5c47119e668867cdb51baff0eca1fdb68/src/lib/getline.cc
/c/j/w/prod/buildsinglereference/.conan/data/gperf/3.1/_/_/build/0a420ff5c47119e668867cdb51baff0eca1fdb68/src/build-aux/compile cl -nologo -MT -O2 -Ob2 -FS -DNDEBUG -D_WIN32_WINNT=_WIN32_WINNT_WIN8 -I/c/j/w/prod/buildsinglereference/.conan/data/gperf/3.1/_/_/build/0a420ff5c47119e668867cdb51baff0eca1fdb68/src/lib -c /c/j/w/prod/buildsinglereference/.conan/data/gperf/3.1/_/_/build/0a420ff5c47119e668867cdb51baff0eca1fdb68/src/lib/hash.cc
```

the problem is that wherever the paths for the .cc files get converted to windows formats, the conversion seems wrong, judging by the error from cl.exe referencing this:
```
C:J:/w/prod/buildsinglereference/.conan/data/gperf/3.1/_/_/build/0a420ff5c47119e668867cdb51baff0eca1fdb68/src/lib/getline.cc
```

I can reproduce this locally. And I run the same commands in order (command prompt + call .bat + msys2 bash + source conanbuild.sh) everything works fine. Will keep testing

**Comment by jcar87** [comment]
The issue is due to an interaction between the `compile` wrapper and the msys2 shell. 
Indeed the compile wrapper has different logic for files with the .cc extension, and already performs the unix->windows path conversion (correctly as far as I can see), and then passes the `-c -TpC:/path/to/file.cc`  when invoking `cl.exe`  (`/Tp` being the C++ mode of cl.exe). 
The problem arises when msys2 detects a path in the command line arguments provided to `cl.exe` and then performs a conversion of its own before invoking the process  - it starts converting anything that follows `-TpC:`, and thinks the path starts at `/j/w/...` , causing the final invocation to look like `-TpC:J:/w/prod/...` as we see above.

The reason this doesn't fail for `.c` files is because the compile wrapper doesnt perform path conversions in this instance and the command line argument isn't prepended with anything. 
The reason this works currently as the recipe it is in master is because `./configure` is called "in-place" in the source directory and all the paths in the makefile are relative, whereas now they are absolute - which causes `compile` to perform path conversions.

An easy way out is to tell msys2 to _NOT_ perform path conversions for the `-Tp` argument:
```
env.define("MSYS2_ARG_CONV_EXCL", "-Tp")

As mentioned in there [docs](https://www.msys2.org/docs/filesystem-paths/), the logic is not perfect for all cases, and looks like we just stumbled upon one of those cases.

File: recipes/gperf/all/conanfile.py [hunk 1]
```diff
@@ -11,88 +16,78 @@ class GperfConan(ConanFile):
...
+            ar_wrapper = unix_path(self, os.path.join(self.source_folder, "build-aux", "ar-lib"))
+            env.define("CC", f"{compile_wrapper} cl -nologo")
+            env.define("CXX", f"{compile_wrapper} cl -nologo")
+            env.append("CPPFLAGS", "-D_WIN32_WINNT=_WIN32_WINNT_WIN8")
+            env.define("LD", "link -nologo")
+            env.define("AR", f"{ar_wrapper} \"lib -nologo\"")
+            env.define("NM", "dumpbin -symbols")
+            env.define("OBJDUMP", ":")
+            env.define("RANLIB", ":")
+            env.define("STRIP", ":")
```
Comment (jcar87) [inline recipes/gperf/all/conanfile.py:70]:
```suggestion
            env.define("STRIP", ":")
            
            #Prevent msys2 from performing erroneous path conversions for C++ files
            # when invoking cl.exe as this is already handled by the compile wrapper.
            env.define("MSYS2_ARG_CONV_EXCL", "-Tp") 
```

**Comment by SpaceIm** [comment]
Thanks @jcar87 

> The reason this works currently as the recipe it is in master is because ./configure is called "in-place" in the source directory and all the paths in the makefile are relative, whereas now they are absolute - which causes compile to perform path conversions.

Is it possible to "fix" the new Autotools helper to always call configure in-place?

**Comment by SpaceIm** [comment]
Awesome

**Comment by jcar87** [comment]
> Thanks @jcar87
> 
> > The reason this works currently as the recipe it is in master is because ./configure is called "in-place" in the source directory and all the paths in the makefile are relative, whereas now they are absolute - which causes compile to perform path conversions.
> 
> Is it possible to "fix" the new Autotools helper to always call configure in-place?

Here @czoido may have better context than me to provide an answer. 

As far as I know, out-of-tree builds should work fine with Autotools simply by invoking the configure script in a different cwd than the source folder. An interesting think is that I think the deciding factor in whether or not the paths in the Makefiles end up being relative or absolute lies in "how" we invoke the configure script. The current implementation of `autotools.configure()` does it by invoking the script in its absolute location, but if it did something like `../src/configure`, I suspect that would be enough for paths in the generated Makefiles to be relative. Worth testing!

**Review by toge — APPROVED** [review]

**Review by uilianries — APPROVED** [review]
LGTM

**Pull request merged**